src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -525,6 +525,135 @@ public override BoundNode VisitTypeOrInstanceInitializers(BoundTypeOrInstanceIni
+                    ImmutableArray.Create<BoundExpression>(
+                        indexAssign,
+                        arrayAssign),
+                    F.Conditional(
+                        F.Property(indexLocal, indexFromEndSymbol),
+                        F.ArrayAccess(arrayLocal, ImmutableArray.Create<BoundExpression>(F.Binary(
+                            BinaryOperatorKind.Subtraction,
+                            F.SpecialType(SpecialType.System_Int32),
+                            F.ArrayLength(arrayLocal),
+                            indexValueExpr))),
+                        F.ArrayAccess(arrayLocal, ImmutableArray.Create(indexValueExpr)),
+                        node.Type));
+            }
+            else if (indexType == _compilation.GetWellKnownType(WellKnownType.System_Range))
+            {
+                // array[Range] is translated to:
+                // var start = range.Start.FromEnd ? array.Length - range.Start.Value : range.Start.Value;
+                // var end = range.End.FromEnd ? array.Length - range.End.Value : range.End.Value;
+                // var length = end - start;
+                // var newArr = new T[length];
+                // Array.Copy(array, start, newArr, 0, length);
+                // push newArray
+
+                var rangeStartSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Range__Start);
+                var rangeEndSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Range__End);
+                var arrayCopySymbol = F.WellKnownMethod(WellKnownMember.System_Array__Copy);
+
+                var startLocal = F.StoreToTemp(
+                    F.Conditional(
+                        F.Property(F.Property(indexLocal, rangeStartSymbol), indexFromEndSymbol),
+                        F.Binary(
+                            BinaryOperatorKind.Subtraction,
+                            F.SpecialType(SpecialType.System_Int32),
+                            F.ArrayLength(arrayLocal),
+                            F.Property(F.Property(indexLocal, rangeEndSymbol), indexValueSymbol)),
+                        F.Property(F.Property(indexLocal, rangeStartSymbol), indexValueSymbol),
+                        F.SpecialType(SpecialType.System_Int32)),
+                    out BoundAssignmentOperator startAssign);
+                var endLocal = F.StoreToTemp(
+                    F.Conditional(
+                        F.Property(F.Property(indexLocal, rangeEndSymbol), indexFromEndSymbol),
+                        F.Binary(
+                            BinaryOperatorKind.Subtraction,
+                            F.SpecialType(SpecialType.System_Int32),
+                            F.ArrayLength(arrayLocal),
+                            F.Property(F.Property(indexLocal, rangeEndSymbol), indexValueSymbol)),
+                        F.Property(F.Property(indexLocal, rangeEndSymbol), indexValueSymbol),
+                        F.SpecialType(SpecialType.System_Int32)),
```

Comment:
Can this be extracted to a helper method, for use in `array[Index]` and in the two places here for `array[Range]`?